### PR TITLE
Cldc 1670 reactivate locations

### PR DESCRIPTION
**Pull request by kosiakkatrina**
Add reactivation page (without a confirmation page) that allows reactivating a location.
Update existing location_deactivation_period records that don't have a reactivation_date to the selected/entered reactivation date.
image
image
Add `reactivating_soon` status
Update availability to only display periods valuable to the user and combine intersecting active periods
image
Extract `active_periods` method to model, so that it could be reused, and use it when constructing the availability timeline.
Use `location_deactivation_periods` table as a log of events/deactivations and stitch the human readable active periods in `active_periods` method.
Add a validation for a deactivation date that is created during a different deactivation, to minimise overlapping deactivations

### CODE REVIEW

File: spec/helpers/locations_helper_spec.rb [hunk 1]
```diff
@@ -67,7 +67,7 @@
     end
 
     context "when viewing availability" do
-      context "with are no deactivations" do
+      context "with ae no deactivations" do
```
Comment (natdeanlewissoftwire) [inline spec/helpers/locations_helper_spec.rb]: I assume this isn't what you meant 😅

Comment (kosiakkatrina) [reply spec/helpers/locations_helper_spec.rb]: 🤦 Thanks!

File: app/controllers/locations_controller.rb [hunk 2]
```diff
@@ -189,22 +202,39 @@ def deactivate_success_notice
...
     @location.lettings_logs.filter_by_before_startdate(deactivation_date.to_time).update!(location: nil, scheme: nil)
   end
 
   def deactivation_date
+    toggle_date("deactivation_date")
+  end
+
+  def reactivation_date
+    toggle_date("reactivation_date")
+  end
```
Comment (kosiakkatrina) [inline app/controllers/locations_controller.rb]: could not extract these

File: app/models/location.rb [hunk 3]
```diff
@@ -399,15 +407,37 @@ def deactivation_date_errors
 
     if deactivation_date.blank?
       if deactivation_date_type.blank?
-        errors.add(:deactivation_date_type, message: I18n.t("validations.location.deactivation_date.not_selected"))
+        errors.add(:deactivation_date_type, message: I18n.t("validations.location.toggle_date.not_selected"))
       elsif deactivation_date_type == "other"
-        errors.add(:deactivation_date, message: I18n.t("validations.location.deactivation_date.invalid"))
+        errors.add(:deactivation_date, message: I18n.t("validations.location.toggle_date.invalid"))
       end
+    elsif location_deactivation_periods.any? { |period| deactivation_date.between?(period.deactivation_date, period.reactivation_date - 1.day) }
```
Comment (kosiakkatrina) [inline app/models/location.rb]: check this using sql

File: app/helpers/locations_helper.rb [hunk 4]
```diff
@@ -42,12 +42,43 @@ def display_location_attributes(location)
...
+      end
     end
-    availability
+    availability.strip
+  end
+
+private
+
+  def remove_overlapping_and_empty_periods(periods)
+    periods.select { |period| ((period.to.nil? && period.from.present?) || (period.from.present? && period.from < period.to)) }
```
Comment (kosiakkatrina) [inline app/helpers/locations_helper.rb]: extract `period.from.present?`

File: app/controllers/locations_controller.rb [hunk 5]
```diff
@@ -41,18 +43,31 @@ def deactivate_confirm
   end
 
   def deactivate
-    @location.run_deactivation_validations!
-
-    if @location.location_deactivation_periods.create!(deactivation_date:) && update_affected_logs
+    if LocationDeactivationPeriod.create!(location_id: @location.id, deactivation_date: params[:deactivation_date]) && update_affected_logs
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]:
```suggestion
    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && update_affected_logs!
```

Two suggestions here:

1. Come through the association so you don't need to be explicit about the location
2. Make the second method a `!` method to show that it changes state.

File: app/controllers/locations_controller.rb [hunk 6]
```diff
@@ -185,26 +200,35 @@ def deactivate_success_notice
...
+    when :active
+      "#{@location.name} has been reactivated"
+    when :reactivating_soon
+      "#{@location.name} will reactivate on #{toggle_date('reactivation_date').to_time.to_formatted_s(:govuk_date)}"
     end
   end
 
   def update_affected_logs
-    @location.lettings_logs.filter_by_before_startdate(deactivation_date.to_time).update!(location: nil, scheme: nil)
+    @location.lettings_logs.filter_by_before_startdate(params[:deactivation_date].to_time).update!(location: nil, scheme: nil)
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb:217]: Or to be explicit... `reset_location_and_scheme_for_logs!` ?

File: app/helpers/locations_helper.rb [hunk 7]
```diff
@@ -42,12 +42,43 @@ def display_location_attributes(location)
...
+  def remove_overlapping_and_empty_periods(periods)
+    periods.select { |period| period.from.present? && (period.to.nil? || period.from < period.to) }
+  end
+
+  def remove_nested_periods(periods)
+    periods.select { |inner_period| periods.none? { |outer_period| is_nested?(inner_period, outer_period) } }
+  end
+
+  def is_nested?(inner, outer)
+    return false if inner == outer || [inner.deactivation_date, inner.reactivation_date, outer.deactivation_date, outer.reactivation_date].any?(&:blank?)
```
Comment (jamesrose) [inline app/helpers/locations_helper.rb]:
```suggestion
    return false if inner == outer
    return false if [inner.deactivation_date, inner.reactivation_date, outer.deactivation_date, outer.reactivation_date].any?(&:blank?)
```

A little easier to parse!

File: app/helpers/locations_helper.rb [hunk 8]
```diff
@@ -42,12 +42,44 @@ def display_location_attributes(location)
...
+    periods.select { |period| period.from.present? && (period.to.nil? || period.from < period.to) }
+  end
+
+  def remove_nested_periods(periods)
+    periods.select { |inner_period| periods.none? { |outer_period| is_nested?(inner_period, outer_period) } }
+  end
+
+  def is_nested?(inner, outer)
+    return false if inner == outer
+    return false if [inner.deactivation_date, inner.reactivation_date, outer.deactivation_date, outer.reactivation_date].any?(&:blank?)
```
Comment (natdeanlewissoftwire) [inline app/helpers/locations_helper.rb:81]: could this be an `or`?

File: app/helpers/locations_helper.rb [hunk 9]
```diff
@@ -42,12 +42,43 @@ def display_location_attributes(location)
...
+      end
     end
-    availability
+    availability.strip
+  end
+
+private
+
+  def remove_overlapping_and_empty_periods(periods)
+    periods.select { |period| period.from.present? && (period.to.nil? || period.from < period.to) }
```
Comment (jamesrose) [inline app/helpers/locations_helper.rb:72]:
```suggestion
    periods.select { |(from, to)| from.present? && (to.nil? || from < to) }
```

Part 2 if you use my suggestion without structs. Again: not sure if this is actually better.

File: app/helpers/locations_helper.rb [hunk 10]
```diff
@@ -42,12 +42,43 @@ def display_location_attributes(location)
...
+    periods = [ActivePeriod.new(location.available_from, nil)]
+
+    sorted_deactivation_periods = remove_nested_periods(location.location_deactivation_periods.sort_by(&:deactivation_date))
+    sorted_deactivation_periods.each do |deactivation|
+      periods.find { |period| period.to.nil? }.to = deactivation.deactivation_date
+      periods << ActivePeriod.new(deactivation.reactivation_date, nil)
+    end
+
+    remove_overlapping_and_empty_periods(periods)
+  end
```
Comment (jamesrose) [inline app/helpers/locations_helper.rb]:
```suggestion
  ActivePeriod = Struct.new(:from, :to)
  def active_periods(location)
    periods = [ActivePeriod.new(location.available_from, nil)]

    sorted_deactivation_periods = remove_nested_periods(location.location_deactivation_periods.sort_by(&:deactivation_date))
    sorted_deactivation_periods.each do |deactivation|
      periods.last.to = deactivation.deactivation_date
      periods << ActivePeriod.new(deactivation.reactivation_date, nil)
    end

    remove_overlapping_and_empty_periods(periods)
  end
```

I wonder: do we need to check whether the last period `to` is nil?

Here's another way to write it without structs. Whether this is easier to read I'm not convinced!

```suggestion
  def active_periods(location)
    periods = [[location.available_from, nil]]

    sorted_deactivation_periods = remove_nested_periods(location.location_deactivation_periods.sort_by(&:deactivation_date))
    sorted_deactivation_periods.each do |deactivation|
      periods.last[1] = deactivation.deactivation_date
      periods << [deactivation.reactivation_date, nil]
    end

    remove_overlapping_and_empty_periods(periods)
  end
```

File: app/helpers/locations_helper.rb (hunk 8 above)
Comment (kosiakkatrina) [reply app/helpers/locations_helper.rb:81]: https://github.com/communitiesuk/submit-social-housing-lettings-and-sales-data/pull/1007#discussion_r1028311569 👀

File: app/controllers/locations_controller.rb [hunk 11]
```diff
@@ -41,18 +43,31 @@ def deactivate_confirm
   end
 
   def deactivate
-    @location.run_deactivation_validations!
-
-    if @location.location_deactivation_periods.create!(deactivation_date:) && update_affected_logs
+    if @location.location_deactivation_periods.create!(location_id: @location.id, deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
```
Comment (jamesrose) [inline app/controllers/locations_controller.rb]:
```suggestion
    if @location.location_deactivation_periods.create!(deactivation_date: params[:deactivation_date]) && reset_location_and_scheme_for_logs!
```

Can remove `location_id` from here now :)

File: app/controllers/schemes_controller.rb [hunk 12]
```diff
@@ -299,28 +299,28 @@ def deactivate_success_notice
...
-    year = params[:scheme]["deactivation_date(1i)"]
+    day = params[:scheme_deactivation_period]["deactivation_date(3i)"]
+    month = params[:scheme_deactivation_period]["deactivation_date(2i)"]
+    year = params[:scheme_deactivation_period]["deactivation_date(1i)"]
     return nil if [day, month, year].any?(&:blank?)
 
     Time.zone.local(year.to_i, month.to_i, day.to_i) if Date.valid_date?(year.to_i, month.to_i, day.to_i)
   end
 
   def update_affected_logs
```
Comment (jamesrose) [inline app/controllers/schemes_controller.rb]:
```suggestion
  def reset_location_and_scheme_for_logs!
```

Same name here?

File: app/helpers/locations_helper.rb (hunk 10 above)
Comment (kosiakkatrina) [reply app/helpers/locations_helper.rb]: Updated the .last suggestion, but I quite like the struct if there are no strong feelings there 👀

Comment (jamesrose) [reply app/helpers/locations_helper.rb]: None!

**Review by jamesrose — APPROVED** [review]

**Pull request merged**